File: test/lsp-client.py
#!/usr/bin/env python3
"""
Minimal LSP client for testing ocaml-lsp-server.

Usage:
    python3 lsp-client.py <command> [args...]

Commands:
    initialize      Send initialize request
    hover <file> <line> <col>    Request hover info
    completion <file> <line> <col>    Request completions
    format <file>   Request formatting
    shutdown        Send shutdown request
"""

import json
import subprocess
import sys
from pathlib import Path


class LSPClient:
    def __init__(self):
        self.proc = subprocess.Popen(
            ["ocamllsp"],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        self.request_id = 0

    def _send(self, message: dict) -> None:
        content = json.dumps(message)
        header = f"Content-Length: {len(content)}\r\n\r\n"
        self.proc.stdin.write(header.encode())
        self.proc.stdin.write(content.encode())
        self.proc.stdin.flush()

    def _receive(self) -> dict:
        # Read header
        headers = {}
        while True:
            line = self.proc.stdout.readline().decode().strip()
            if not line:
                break
            key, value = line.split(": ", 1)
            headers[key] = value

        # Read content
        content_length = int(headers["Content-Length"])
        content = self.proc.stdout.read(content_length).decode()
        return json.loads(content)

    def request(self, method: str, params: dict = None) -> dict:
        self.request_id += 1
        message = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
        }
        if params:
            message["params"] = params
        self._send(message)
        return self._receive()

    def notify(self, method: str, params: dict = None) -> None:
        message = {
            "jsonrpc": "2.0",
            "method": method,
        }
        if params:
            message["params"] = params
        self._send(message)

    def initialize(self, root_path: str = ".") -> dict:
        root_uri = Path(root_path).absolute().as_uri()
        return self.request("initialize", {
            "processId": None,
            "rootUri": root_uri,
            "capabilities": {
                "textDocument": {
                    "hover": {"contentFormat": ["markdown", "plaintext"]},
                    "completion": {
                        "completionItem": {"snippetSupport": False}
                    },
                    "formatting": {},
                }
            },
        })

    def initialized(self) -> None:
        self.notify("initialized", {})

    def did_open(self, file_path: str, content: str = None) -> None:
        path = Path(file_path).absolute()
        if content is None:
            content = path.read_text()
        self.notify("textDocument/didOpen", {
            "textDocument": {
                "uri": path.as_uri(),
                "languageId": "ocaml",
                "version": 1,
                "text": content,
            }
        })

    def hover(self, file_path: str, line: int, col: int) -> dict:
        path = Path(file_path).absolute()
        return self.request("textDocument/hover", {
            "textDocument": {"uri": path.as_uri()},
            "position": {"line": line, "character": col},
        })

    def completion(self, file_path: str, line: int, col: int) -> dict:
        path = Path(file_path).absolute()
        return self.request("textDocument/completion", {
            "textDocument": {"uri": path.as_uri()},
            "position": {"line": line, "character": col},
        })

    def formatting(self, file_path: str) -> dict:
        path = Path(file_path).absolute()
        return self.request("textDocument/formatting", {
            "textDocument": {"uri": path.as_uri()},
            "options": {"tabSize": 2, "insertSpaces": True},
        })

    def shutdown(self) -> dict:
        return self.request("shutdown")

    def exit(self) -> None:
        self.notify("exit")
        self.proc.wait()


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    command = sys.argv[1]
    client = LSPClient()

    try:
        if command == "initialize":
            root = sys.argv[2] if len(sys.argv) > 2 else "."
            result = client.initialize(root)
            print(json.dumps(result, indent=2))
            client.initialized()

        elif command == "hover":
            if len(sys.argv) < 5:
                print("Usage: lsp-client.py hover <file> <line> <col>")
                sys.exit(1)
            client.initialize()
            client.initialized()
            file_path, line, col = sys.argv[2], int(sys.argv[3]), int(sys.argv[4])
            client.did_open(file_path)
            result = client.hover(file_path, line, col)
            print(json.dumps(result, indent=2))

        elif command == "completion":
            if len(sys.argv) < 5:
                print("Usage: lsp-client.py completion <file> <line> <col>")
                sys.exit(1)
            client.initialize()
            client.initialized()
            file_path, line, col = sys.argv[2], int(sys.argv[3]), int(sys.argv[4])
            client.did_open(file_path)
            result = client.completion(file_path, line, col)
            print(json.dumps(result, indent=2))

        elif command == "format":
            if len(sys.argv) < 3:
                print("Usage: lsp-client.py format <file>")
                sys.exit(1)
            client.initialize()
            client.initialized()
            file_path = sys.argv[2]
            client.did_open(file_path)
            result = client.formatting(file_path)
            print(json.dumps(result, indent=2))

        elif command == "shutdown":
            client.initialize()
            client.initialized()
            result = client.shutdown()
            print(json.dumps(result, indent=2))
            client.exit()

        else:
            print(f"Unknown command: {command}")
            print(__doc__)
            sys.exit(1)

    finally:
        if client.proc.poll() is None:
            client.proc.terminate()


if __name__ == "__main__":
    main()

File: test/lsp-client/lsp_client.ml
let request_id = ref 0

(* JSON-RPC 2.0 with Content-Length framing *)

let send oc (msg : Yojson.Safe.t) =
  let content = Yojson.Safe.to_string msg in
  let header = Printf.sprintf "Content-Length: %d\r\n\r\n" (String.length content) in
  output_string oc header;
  output_string oc content;
  flush oc

let read_headers ic =
  let headers = Hashtbl.create 4 in
  let rec loop () =
    let line = input_line ic in
    let line = String.trim line in
    if line = "" then headers
    else begin
      (match String.split_on_char ':' line with
       | key :: rest ->
         let value = String.trim (String.concat ":" rest) in
         Hashtbl.replace headers key value
       | _ -> ());
      loop ()
    end
  in
  loop ()

let receive ic =
  let headers = read_headers ic in
  let content_length =
    int_of_string (Hashtbl.find headers "Content-Length")
  in
  let buf = Bytes.create content_length in
  really_input ic buf 0 content_length;
  Yojson.Safe.from_string (Bytes.to_string buf)

(* Skip server-initiated notifications (no "id" field) and return the
   next response that carries an "id". This prevents flaky tests when
   ocamllsp sends unsolicited notifications such as publishDiagnostics
   before replying to our request. *)
let receive_response ic =
  let rec loop () =
    let msg = receive ic in
    match Yojson.Safe.Util.member "id" msg with
    | `Null -> loop ()
    | _ -> msg
  in
  loop ()

let request oc ic method_ params =
  incr request_id;
  let msg =
    `Assoc
      ([ ("jsonrpc", `String "2.0");
         ("id", `Int !request_id);
         ("method", `String method_) ]
       @ (match params with
          | Some p -> [ ("params", p) ]
          | None -> []))
  in
  send oc msg;
  receive_response ic

let notify oc method_ params =
  let msg =
    `Assoc
      ([ ("jsonrpc", `String "2.0");
         ("method", `String method_) ]
       @ (match params with
          | Some p -> [ ("params", p) ]
          | None -> []))
  in
  send oc msg

(* URI helpers *)

let file_uri path =
  let abs =
    if Filename.is_relative path then Filename.concat (Sys.getcwd ()) path
    else path
  in
  "file://" ^ abs

(* LSP commands *)

let initialize oc ic root_path =
  let root_uri = file_uri root_path in
  let params =
    `Assoc
      [ ("processId", `Null);
        ("rootUri", `String root_uri);
        ("capabilities",
         `Assoc
           [ ("textDocument",
              `Assoc
                [ ("hover",
                   `Assoc
                     [ ("contentFormat",
                        `List [ `String "markdown"; `String "plaintext" ]) ]);
                  ("completion",
                   `Assoc
                     [ ("completionItem",
                        `Assoc [ ("snippetSupport", `Bool false) ]) ]);
                  ("formatting", `Assoc []) ]) ]) ]
  in
  let result = request oc ic "initialize" (Some params) in
  notify oc "initialized" (Some (`Assoc []));
  result

let did_open oc file_path =
  let uri = file_uri file_path in
  let content =
    let ch = open_in file_path in
    Fun.protect
      ~finally:(fun () -> close_in ch)
      (fun () ->
        let len = in_channel_length ch in
        really_input_string ch len)
  in
  notify oc "textDocument/didOpen"
    (Some
       (`Assoc
          [ ("textDocument",
             `Assoc
               [ ("uri", `String uri);
                 ("languageId", `String "ocaml");
                 ("version", `Int 1);
                 ("text", `String content) ]) ]))

let hover oc ic file_path line col =
  let uri = file_uri file_path in
  request oc ic "textDocument/hover"
    (Some
       (`Assoc
          [ ("textDocument", `Assoc [ ("uri", `String uri) ]);
            ("position",
             `Assoc [ ("line", `Int line); ("character", `Int col) ]) ]))

let completion oc ic file_path line col =
  let uri = file_uri file_path in
  request oc ic "textDocument/completion"
    (Some
       (`Assoc
          [ ("textDocument", `Assoc [ ("uri", `String uri) ]);
            ("position",
             `Assoc [ ("line", `Int line); ("character", `Int col) ]) ]))

let formatting oc ic file_path =
  let uri = file_uri file_path in
  request oc ic "textDocument/formatting"
    (Some
       (`Assoc
          [ ("textDocument", `Assoc [ ("uri", `String uri) ]);
            ("options",
             `Assoc
               [ ("tabSize", `Int 2); ("insertSpaces", `Bool true) ]) ]))

let shutdown oc ic =
  let result = request oc ic "shutdown" None in
  notify oc "exit" None;
  result

(* Subprocess management *)

let with_lsp_server f =
  let in_r, in_w = Unix.pipe () in
  let out_r, out_w = Unix.pipe () in
  let pid =
    Unix.create_process "ocamllsp" [| "ocamllsp" |] in_r out_w Unix.stderr
  in
  Unix.close in_r;
  Unix.close out_w;
  let oc = Unix.out_channel_of_descr in_w in
  let ic = Unix.in_channel_of_descr out_r in
  Fun.protect
    ~finally:(fun () ->
      (try close_out oc with _ -> ());
      (try close_in ic with _ -> ());
      (try Unix.kill pid Sys.sigterm with _ -> ());
      ignore (Unix.waitpid [] pid))
    (fun () -> f oc ic)

let () =
  if Array.length Sys.argv < 2 then begin
    Printf.eprintf
      "Usage: lsp_client <command> [args...]\n\
       Commands:\n\
      \  initialize [root]           Initialize LSP server\n\
      \  hover <file> <line> <col>   Request hover info\n\
      \  completion <file> <line> <col>  Request completions\n\
      \  format <file>               Request formatting\n\
      \  shutdown                    Send shutdown request\n";
    exit 1
  end;
  let command = Sys.argv.(1) in
  with_lsp_server (fun oc ic ->
    match command with
    | "initialize" ->
      let root = if Array.length Sys.argv > 2 then Sys.argv.(2) else "." in
      let result = initialize oc ic root in
      print_endline (Yojson.Safe.pretty_to_string result)
    | "hover" ->
      if Array.length Sys.argv < 5 then begin
        Printf.eprintf "Usage: lsp_client hover <file> <line> <col>\n";
        exit 1
      end;
      ignore (initialize oc ic ".");
      let file_path = Sys.argv.(2) in
      let line = int_of_string Sys.argv.(3) in
      let col = int_of_string Sys.argv.(4) in
      did_open oc file_path;
      let result = hover oc ic file_path line col in
      print_endline (Yojson.Safe.pretty_to_string result)
    | "completion" ->
      if Array.length Sys.argv < 5 then begin
        Printf.eprintf "Usage: lsp_client completion <file> <line> <col>\n";
        exit 1
      end;
      ignore (initialize oc ic ".");
      let file_path = Sys.argv.(2) in
      let line = int_of_string Sys.argv.(3) in
      let col = int_of_string Sys.argv.(4) in
      did_open oc file_path;
      let result = completion oc ic file_path line col in
      print_endline (Yojson.Safe.pretty_to_string result)
    | "format" ->
      if Array.length Sys.argv < 3 then begin
        Printf.eprintf "Usage: lsp_client format <file>\n";
        exit 1
      end;
      ignore (initialize oc ic ".");
      let file_path = Sys.argv.(2) in
      did_open oc file_path;
      let result = formatting oc ic file_path in
      print_endline (Yojson.Safe.pretty_to_string result)
    | "shutdown" ->
      ignore (initialize oc ic ".");
      let result = shutdown oc ic in
      print_endline (Yojson.Safe.pretty_to_string result)
    | _ ->
      Printf.eprintf "Unknown command: %s\n" command;
      exit 1)
